subsys/fs/zffs/src/zffs_misc.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2018 Findlay Feng
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "zffs/misc.h"
+#include "zffs/dir.h"
+#include "zffs/file.h"
+#include "zffs/object.h"
+#include "zffs/path.h"
+#include "zffs/tree.h"
+
+void zffs_misc_lock(struct zffs_data *zffs)
+{
+	k_mutex_lock(&zffs->lock, K_FOREVER);
+}
+
+void zffs_misc_unlock(struct zffs_data *zffs)
+{
+	k_mutex_unlock(&zffs->lock);
+}
+
+u32_t zffs_misc_get_id(struct zffs_data *zffs)
+{
+	int lock = k_mutex_lock(&zffs->lock, K_NO_WAIT);
+	int id;
+
+	if (zffs->next_id == ZFFS_ROOT_ID) {
+		zffs->next_id += ZFFS_CONFIG_MISC_ID_STEP;
+	}
+
+	id = zffs->next_id;
+	zffs->next_id += ZFFS_CONFIG_MISC_ID_STEP;
+
+	if (!lock) {
+		k_mutex_unlock(&zffs->lock);
+	}
+
+	return id;
+}
+
+u32_t zffs_misc_next_id(struct zffs_data *zffs, u32_t *id)
+{
+	if ((*id + 1) % ZFFS_CONFIG_MISC_ID_STEP == 0) {
+		*id = zffs_misc_get_id(zffs);
+	}
+	else {
+		(*id)++;
+	}
+
+	return *id;
+}
+
+int zffs_misc_restore(struct zffs_data *zffs)
+{
+	struct zffs_area_pointer pointer = zffs_data_pointer(zffs);
+	u32_t id, addr;
+	u32_t key_count, key_max;
+	struct zffs_node_data node_data;
+	int rc;
+
+	zffs->data_write_addr = 0;
+	zffs->swap_write_addr = 0;
+
+	rc = zffs_tree_init(zffs);
+	if (rc) {
+		return rc;
+	}
+
+	rc = zffs_tree_info(zffs, &key_count, &key_max, NULL,
+			     &zffs->data_write_addr);
+	if (rc) {
+		return rc;
+	}
+
+	if (key_count > 0) {
+		zffs_area_addr_to_pointer(zffs, zffs->data_write_addr,
+					   &pointer);
+		rc = zffs_object_check(zffs, &pointer, &id);
+		if (rc) {
+			// 删除这个键 地址块被遗弃
```

Comment:
Commenting in english (and with /**/, not //) would be better for all the contributors ;)

Reply:
This is a todo reserved for myself.
It will disappear when it is finished